Autosave drafts after delay

aduth opened the pull request:
Partial resolution for #1773

This pull request seeks to implement an incomplete autosave behavior, currently applied to drafts only. Published posts lack API endpoint requirements for creating autosaves (see https://github.com/WordPress/gutenberg/issues/1773#issuecomment-316434091) and will need to be implemented separately. Further improvements may be needed for handling autosave for private or scheduled posts, currently ignored as of these changes.

__Implementation notes:__

Autosave occurs on a 10 second debounce. The current post editor [hooks to heartbeat lock refresh](https://github.com/WordPress/WordPress/blob/a798266/wp-admin/js/post.js#L178-L213) to trigger an autosave. More consideration may be needed for autosave frequency and use of browser storage for locally saved copies of the post. Discussion may be better reserved for #1773, though this pull request seeks to start exploring a minimal solution.

__Testing instructions:__

Verify that autosave occurs for a draft 10 seconds after a change:

1. Navigate to Gutenberg > New Post
2. Enter a title
3. Wait 10 seconds
4. Note that the draft saves

aduth committed:
Rename isEditedPostPublished to isCurrentPostPublished

"edited post" in context of selectors should refer to current post + edited values. Current usage tests against saved post only. Needed to distinguish for "isEditedPostPublished" testing current post with edited values applied.

aduth committed:
Autosave drafts after debounce delay

westonruter commented:
Autosave is also closely related to the need to previewing changes on the frontend: #1673.

youknowriad commented on editor/effects.js:
```diff
@@ -210,4 +221,28 @@ export default {
+	REPLACE_BLOCKS: () => queueAutosave(),
+	REMOVE_BLOCKS: () => queueAutosave(),
+	EDIT_POST: () => queueAutosave(),
+	MARK_DIRTY: () => queueAutosave(),
```
Why do we need all these effects? I suspect dirtiness? Can't we use an internal somewhere instead?

aduth replied:
>Why do we need all these effects? I suspect dirtiness? Can't we use an internal somewhere instead?

Yeah, all the potentially dirty-ing effects, see also: reducer for `editor.dirty`.

As remarked in #1967, I'd like to look to reimplementing dirtiness as a diff against the original saved post, in which case we could reimplement this as a subscribe on the store (i.e. `if ( currentDirty !== lastDirty ) queueAutosave()`).

youknowriad commented on editor/selectors.js:
```diff
@@ -168,19 +168,39 @@ export function getEditedPostVisibility( state ) {
+ * @param  {Object}   state Global application state
+ * @return {Boolearn}       Whether the post has been bublished
+ */
+export function isEditedPostPublished( state ) {
```
I'm having trouble understanding if this is really necessary because the status is changed to 'publish', 'private' only when saving. so no real difference with `isCurrentPostPublished`. Btw, good renaming.

aduth replied:
If I recall correctly, the issue is that if the user chooses a date in the past for their draft post, the edited representation of the post becomes "published", but we can't necessarily use the published autosave flow because the saved post is still a draft.

... I think there's still something amiss here though, because in the `AUTOSAVE` effect above I'm still testing against the edited version of the post, not the "current" (saved) copy.

youknowriad commented on editor/effects.js:
```diff
@@ -210,4 +221,28 @@ export default {
+		}
+
+		if ( isEditedPostPublished( state ) ) {
+			// TODO: Publish autosave
```
What's the desired behaviour here?

aduth replied:
>What's the desired behaviour here?

Published posts have a distinct auto-save behavior which is not currently possible to implement due to lacking API support: https://github.com/WordPress/gutenberg/issues/1773#issuecomment-316434091

mtias reviewed (approved):
Works well for me.

aduth commented:
I'm trying to sort through what should be the autosave behavior, particularly around transitioning status. The obvious ones are draft staying draft, and published staying published, but it becomes less clear when the post status is modified in the same session:

Original Status|Status Pending Save|Autosave Behavior
---|---|---
Draft|Draft|Save draft
Published|Published|Autosave (Not Yet Implemented)
Draft|Published|??? (Nothing?)
Published|Draft|??? (Nothing?)

youknowriad commented:
- For the 4th, I'd say auto-save as Draft but keep the "status pending save" in the edits.
- The 5th is more confusing but I don't think we have this possibility in the UI (currently).

Do you know how does Calypso resolve this?

mtias commented:
In general, the changes of status in Calypso performs the action. That's why we have "revert to draft" as a button, to try to keep it binary.

@jasmussen we should look again at the status area from the point of view of published/draft. Related: https://github.com/WordPress/gutenberg/issues/1726

aduth committed:
Set status from auto-draft to draft before save

aduth commented:
Rebased and force-pushed with some new tests. Also updated the condition for handling published autosaves to checking the current (last saved) post in deciding whether to abort. When auto-saving a post which had originally been published, we may need some additional logic, but this can be implemented separately. In the meantime I've included some thoughts as inline comments.

I'd have liked to have been more strict with how autosave behaves depending on the actual data of the edited post. For example, while not possible by user flows, if a draft is applied an unsaved edit of `status: 'publish'`, an auto-save could occur which publishes the post automatically. Again, this won't occur with any current user flows, but seems like it'd be nice to capture at the autosave behavior nonetheless.
